hoverkite-protocol/src/response.rs

```diff
@@ -89,178 +130,157 @@ impl Report {
-                })
-            } else {
-                None
-            }),
-            Some(b'C') | Some(b'c') => Ok(if buffer.len() >= 2 {
-                let side = if buffer.pop_front().unwrap() == b'C' {
-                    side
-                } else {
-                    side.opposite()
-                };
-                let byte = buffer.pop_front().unwrap();
-                let charger_connected = match byte {
-                    b'0' => false,
-                    b'1' => true,
-                    _ => return Err(UnexpectedResponse(byte)),
-                };
-                Some(Self {
-                    side,
-                    response: Response::Report(Report::ChargeState { charger_connected }),
-                })
-            } else {
-                None
-            }),
-            Some(r) => {
-                buffer.pop_front();
-                Err(UnexpectedResponse(r))
-            }
-            None => Ok(None),
+    #[cfg(feature = "std")]
+    pub fn write_to_std(&self, writer: impl std::io::Write) -> std::io::Result<()> {
+        self.write_to(&mut WriteCompat(writer))
+    }
+
+    pub fn write_to<W>(&self, writer: &mut W) -> Result<(), W::Error>
+    where
+        W: embedded_hal::blocking::serial::Write<u8>,
+    {
+        writer.bwrite_all(&[self.side.to_byte()])?;
+        self.response.write_to(writer)
+    }
+
+    pub fn parse(buffer: &[u8]) -> nb::Result<Self, ProtocolError> {
+        if let [side, ref rest @ ..] = *buffer {
+            Ok(SideResponse {
+                side: Side::parse(side)?,
+                response: Response::parse(rest)?,
+            })
+        } else {
+            Err(WouldBlock)
         }
     }
 }
 
-#[cfg(feature = "std")]
 #[cfg(test)]
 mod tests {
     use super::*;
     use test_case::test_case;
 
+    mod log {
+        use super::super::*;
+        use test_case::test_case;
+
+        #[test_case("$" ; "width 1")]
+        #[test_case("¢" ; "width 2")]
+        #[test_case("€" ; "width 3")]
+        #[test_case("𐍈" ; "width 4")]
+        fn too_long_with_unicode_widths(c: &str) {
+            let response = Response::log_from_fmt(format_args!("{}", c.repeat(500)));
+            let log = match response {
+                Response::Log(log) => log,
+                _ => panic!(),
+            };
+            assert!(&log[..].ends_with("..."))
+        }
+
+        #[test]
+        fn parse_too_long() {
+            let buf = format!("\"{}\n", "n".repeat(500));
+            let response = Response::parse(buf.as_bytes());
+            assert_eq!(response, Err(Other(ProtocolError::MessageTooLong)));
+        }
+    }
+
     #[test]
     fn parse_invalid() {
-        let mut buffer = VecDeque::new();
+        let mut buffer = Vec::new();
         buffer.extend(b"x");
         assert_eq!(
-            SideResponse::parse(&mut buffer, Side::Right),
-            Err(UnexpectedResponse(b'x'))
+            SideResponse::parse(&mut buffer),
```

Comment:
Doesn't need to be `&mut`. In fact, this can now just be `SideResponse::parse(b"x")`.